src/annot/pos.rs

```diff
@@ -0,0 +1,427 @@
+
+use annot::contig::Contig;
+use annot::loc::Loc;
+use annot::*;
+use strand::*;
+
+/// Position on a particular, named sequence (e.g. a chromosome).
+///
+/// Parameterized over the type of the reference sequence identifier
+/// and over the strandedness of the position.
+///
+/// The display format for a `Pos` is _chr:pos(+/-)_. A stranded
+/// position must have a _(+)_ or a _(-)_, while an unstranded
+/// position does not.
+///
+/// ```
+/// # use bio_types::annot::ParseAnnotError;
+/// # fn try_main() -> Result<(), Box<ParseAnnotError>> {
+/// use bio_types::annot::pos::Pos;
+/// use bio_types::strand::ReqStrand;
+/// let start = Pos::new("chrIV".to_owned(), 683946, ReqStrand::Reverse);
+/// let start_str = start.to_string();
+/// assert_eq!(start_str, "chrIV:683946(-)");
+/// let start_str_pos = start_str.parse()?;
+/// assert_eq!(start, start_str_pos);
+/// # Ok(())
+/// # }
+/// # fn main() { try_main().unwrap(); }
+/// ```
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct Pos<R, S> {
+    refid: R,
+    pos: isize,
+    strand: S,
+}
+
+impl<R, S> Pos<R, S> {
+    /// Construct a new sequence position
+    ///
+    /// ```
+    /// use std::rc::Rc;
+    /// use bio_types::annot::pos::Pos;
+    /// use bio_types::strand::ReqStrand;
+    /// let chr = Rc::new("chrIV".to_owned());
+    /// let start = Pos::new(chr, 683946, ReqStrand::Reverse);
+    /// ```
+    pub fn new(refid: R, pos: isize, strand: S) -> Self {
+        Pos {
+            refid: refid,
+            pos: pos,
+            strand: strand,
+        }
+    }
+
+    /// Position on the reference sequence (0-based).
+    pub fn pos(&self) -> isize {
+        self.pos
+    }
+
+    /// Convert into an unstranded sequence position
+    pub fn into_unstranded(self) -> Pos<R, NoStrand> {
+        Pos {
+            refid: self.refid,
+            pos: self.pos,
+            strand: NoStrand::Unknown,
+        }
+    }
+
+    /// Convert into a stranded sequence position on the specified strand
+    pub fn into_stranded(self, strand: ReqStrand) -> Pos<R, ReqStrand> {
+        Pos {
+            refid: self.refid,
+            pos: self.pos,
+            strand: strand,
+        }
+    }
+}
+
+impl<R> Pos<R, ReqStrand> {
+    /// Slide the reference position by an offset on the strand of the
+    /// annotation.
+    ///
+    /// # Arguments
+    ///
+    /// * `dist` specifies the offset for sliding the position. A
+    /// positive `dist` will numerically increase the position for
+    /// forward-strand features and decrease it for reverse-strand
+    /// features.
+    ///
+    /// ```
+    /// use bio_types::annot::pos::Pos;
+    /// use bio_types::strand::ReqStrand;
+    /// let mut start = Pos::new("chrIV".to_owned(), 683946, ReqStrand::Reverse);
+    /// assert_eq!(start.to_string(), "chrIV:683946(-)");
+    /// start.slide_on_strand(100);
+    /// assert_eq!(start.to_string(), "chrIV:683846(-)");
+    /// ```
+    pub fn slide_on_strand(&mut self, dist: isize) {
```

Comment:
Maybe just handle this via implementing the `Add` trait, and have different behavior depending on the type of strand.

Reply:
Add returns a new object, when the reference ID isn't Copy (e.g., when it's an owned string) it's nice to have a mutation in place rather than something requiring a clone.

Reply:
I realized that AddAssign (+=) and SubAssign (-=) have the right mutation semantics and used these instead.